plugins/modules/api.py

```diff
@@ -334,37 +334,33 @@ def __init__(self):
         self.update = self.module.params['update']
         self.arbitrary = self.module.params['cmd']
 
-        self.where = None
+        self.result = dict(
+            message=[])
+
+        self.where_list = None
         self.query = self.module.params['query']
         if self.query:
             where_index = self.query.find(' WHERE ')
             if where_index < 0:
                 self.query = self.split_params(self.query)
             else:
+                self.where_list = []
                 where = self.query[where_index + len(' WHERE '):]
+                # create a list of WHERE arguments 
+                for wl in where.split(','):
```

Comment:
Hmm, this is a bit complicated. A comma could show up in a value. For example if `WHERE foo == "bar, baz"`, it would split this into two elements `foo == "bar` and `baz`, and this will result in an error.

Reply:
agree probably ',' can be changed to 'AND' ?

Reply:
It will have the same problem: `WHERE foo == "bar AND baz"` :) I'm afraid we have to properly parse this, instead of simply doing a split with simple string operations.

Reply:
Maybe it would be better to allow to pass the where clauses in as structural data, so we don't have to parse it? Something like:
```
  where:
    - what: foo
      op: '=='
      value: bar AND baz
    - what: something else
      op: '<'
      value: 4
```

Reply:
Is not bad idea however we should structure  it simple as possible 
may be something like 
```
query:
  items: 
    - .id
    - address
  where:
    address:
      op: "=="
      is: 
       - "1.1.1.1"
       - "2.2.2.2"
    network:
      op: "!="
      is: 
       - "255.255.255.0"
    Or:
      address:
        op: "=="
        is: 
         - "3.3.3.3"
         - "4.4.4.4"
```
It will be more easy to be coded, however Or: "is in" where() and not sure how will be more proper to be as yml input (dict structure)
To be key of "where:" or key of "query:"

https://librouteros.readthedocs.io/en/latest/query.html#advanced-usage

Also I'm not sure if case with same key multiple values is valid query at all :|
If I not mistake librouteros.query is https://wiki.mikrotik.com/wiki/Manual:API#Queries

Reply:
```
query:
  items:
    - .id
    - network
    - address
  where:
    - network:
       is: "192.168.58.0"
    - address:
       not: "192.168.58.254/24"
  or:
    - address:
      not: "192.168.58.1/24"
```
? I think looks good (not sure if multiple where: and or: are good combination but will test it ) will fix the code

Reply:
I change the code with the new structure (not completed), however:
If we have same conditions keys with different values we should us librouteros.where Or / In 
For example 
```
        items:
          - .id
          - network
          - address
        where:
          - network:
             is: "192.168.58.0"
          - address:
             is: "192.168.58.254/24"
          - address:
             is: "192.168.58.1/24"
```

This will not return result since (a == 1 and a == 2 ) 'which a ?'  in this case we should use or/in (a == 1 or a == 2 ) - first match win

For reference
```
        items:
          - .id
          - network
          - address
        where:
          - network:
             is: "192.168.58.0"
          - address:
             not: "192.168.58.254/24"
          - address:
             not: "192.168.58.1/24"
```
This will return all other ip addresses which are NOT , for example (1,2,3,4 in a) (a != 1, a !=3) will return (2,4).
Also if (a == 1, a != 3) will return only (1) etc.

What I'm not sure (specially for OR).
If we have (1,2,3,4) (a > 2 or a > 3), this must return (3,4) if OR is valid not only for '=='
This is important since we can change yml to 

 ```
        items:
          - .id
          - network
          - address
        where:
          - network:
             is: "192.168.58.0"
          - address:
             is: 
               - "192.168.58.254/24"
               - "192.168.58.1/24"
```
if type('is') is list:
  ... use or in query

However if we should mix for example '<, >, ==, !=' the yml must be something like (ignore ip values this is a structure  example)
```
        items:
          - .id
          - network
          - address
        where:
          - network:
             is: "192.168.58.0"
          - address:
             or: 
               more: "192.168.58.254/24"
               less: "192.168.58.1/24"
               is: "192.168.58.2/24"
```

Reply:
IMO the new structure should be something that can be validated with the argument spec. (And I would really avoid putting data into dictionary keys if possible, so instead of `<operation>: <value>` it should be something like `operation: <operation>` / `value: <value>`. That makes it easier to process and easier to automatically generate IMO.)

How about limiting it to a or/and structure, so you can have a list of "or" conditions, each containing a list of "and" conditions?